### PR TITLE
[AIRFLOW-1039] Raise AirflowError instead IntegrityError

### PR DESCRIPTION
**Pull request by matusvalo**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1039

### Description
When Two concurrent processes are trying to trigger the same dag with the same execution date at the same time, the IntegrityError is thrown by SQLAlchemy:

IntegrityError: (_mysql_exceptions.IntegrityError) (1062, "Duplicate entry 'poc_dag2-2017-03-24 07:21:22.000000' for key 'dag_id'") [SQL: u'INSERT INTO dag_run (dag_id, execution_date, start_date, end_date, state, run_id, external_trigger, conf) VALUES (%s, %s, now(), %s, %s, %s, %s, %s)'] [parameters: ('poc_dag2', datetime.datetime(2017, 3, 24, 7, 21, 22), None, u'running', '284711d4-109d-11e7-a4aa-005056bc55dd', 1, '\x80\x02}q\x01X\x0b\x00\x00\x00input_files]q\x02X>\x00\x00\x00/homeb/d56032/dev/airflowtest/input2/data15:21:21.845102000.gzq\x03as.')]

This is not consistent with AirflowException returned by trigger_dag() function. Moreover, the session is not rolled back, hence also another exception is occurring:

InvalidRequestError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (_mysql_exceptions.IntegrityError) (1062, "Duplicate entry 'poc_dag2-2017-03-24 06:35:15.000000' for key 'dag_id'") [SQL: u'INSERT INTO dag_run (dag_id, execution_date, start_date, end_date, state, run_id, external_trigger, conf) VALUES (%s, %s, now(), %s, %s, %s, %s, %s)'] [parameters: ('poc_dag2', datetime.datetime(2017, 3, 24, 6, 35, 15), None, u'running', 'b6f3e9ae-1096-11e7-9cea-005056bc55dd', 1, '\x80\x02}q\x01X\x0b\x00\x00\x00input_files]q\x02X>\x00\x00\x00/homeb/d56032/dev/airflowtest/input2/data14:35:13.068563100.gzq\x03as.')]

### CODE REVIEW

**Comment by bolkedebruin** [comment]
-1, sorry this should be handled by locking and checking for existence in the db. Never propagate a db error to the user when it should be handled by the program itself.

**Comment by matusvalo** [comment]
Hi,

I agree that explicit locking and checking is better approach.  But unfortunately, I haven't found any portable way how to lock table before updating it. MySQL has LOCK TABLES statement but this is not portable solution. Moreover the locking should be dag_id-based, because otherwise creating new dag_runs will be serialized (please correct me if I am wrong).

I would like to fix this issue, but I don't know currently how it can be done properly using locking. Do you have any idea?

**Comment by saguziel** [comment]
http://docs.sqlalchemy.org/en/latest/orm/query.html#sqlalchemy.orm.query.Query.with_for_update

**Comment by saguziel** [comment]
See the lock_for_update flag in TaskInstance

**Comment by matusvalo** [comment]
As long as I understand SELECT ... FOR UPDATE and SELECT ... LOCK IN SHARE MODE correctly, they are locking just rows that are selected. 

E.g. suppose that first process runs the following code (trigger_dag.py) and DagRun.find() returns empty list:
```python
    dr = DagRun.find(dag_id=dag_id, run_id=run_id)
    if dr:
        raise AirflowException("Run id {} already exists for dag id {}".format(
            run_id,
            dag_id
        ))
```
It means that there are no dag runs for given run_id, so it will continue with adding new records. Even if SELECT ... FOR UPDATE lock is in place, the second process will  continue and hence it will also try to add new record to table => the second process will fail with exception.

That means that this approach will not protect the use case when two concurrent processes are trying to insert new records to the DB.

File: airflow/models.py [hunk 1]
```diff
@@ -3389,7 +3390,14 @@ def create_dagrun(self,
             state=state
         )
         session.add(run)
-        session.commit()
+        try:
```
Comment (gwax) [inline airflow/models.py:3393]: In addition, please query the database to ask if a matching dag run already exists.

**Comment by ashb** [comment]
Old issue, with merge conflicts. Closing this as a stale PR.

**Pull request closed without merging**